### PR TITLE
o/snapstate: implement transactional flag

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Implement the transactional flag that was introduced in the API by
previous commits, when installing and refreshing snaps.

This is the third part of the original PR https://github.com/snapcore/snapd/pull/11255

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -1218,13 +1218,16 @@ func InstallMany(st *state.State, names []string, userID int, flags *Flags) ([]s
...
+		transLane = st.NewLane()
+	}
 	tasksets := make([]*state.TaskSet, 0, len(installs))
 	for _, sar := range installs {
 		info := sar.Info
 		var snapst SnapState
-		var flags Flags
 
-		flags, err := ensureInstallPreconditions(st, info, flags, &snapst)
+		flags, err := ensureInstallPreconditions(st, info, *flags, &snapst)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1230]: this change doesn't look quite right, as the result will be the input to the next ensureInstallPreconditions, instead the input should always be the original value. We might need a test to prove the difference

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1218,13 +1218,16 @@ func InstallMany(st *state.State, names []string, userID int, flags *Flags) ([]s
 		return nil, nil, err
 	}
 
+	var transLane int
+	if flags.Transactional {
+		transLane = st.NewLane()
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1223]: I would prefer: transactionLane

**Review by pedronis — CHANGES_REQUESTED** [review]
did a first quick pass

File: overlord/snapstate/snapstate.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go]: Oh, right, I overlooked this was in a per-snap loop. Changed now.

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go]: Done

**Review by MiguelPires — COMMENTED** [review]
Just a few test related comments and nitpicks

File: tests/main/refresh-many-transactional/task.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    fi
+
+    echo "Sanity check for the fake store"
+    snap refresh 2>&1 | MATCH "All snaps up to date."
+
+    echo "When the store is configured to make them refreshable"
+    #shellcheck source=tests/lib/store.sh
+    . "$TESTSLIB"/store.sh
+    init_fake_refreshes "$BLOB_DIR" test-snapd-tools
+    retry -n 4 --wait 0.5 test -e "$BLOB_DIR"/test-snapd-tools*fake1*.snap
```
Comment (bboozzoo) [inline tests/main/refresh-many-transactional/task.yaml:55]: Same

File: tests/main/refresh-many-transactional/task.yaml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    init_fake_refreshes  "$BLOB_DIR" test-snapd-python-webserver
+    retry -n 4 --wait 0.5 test -e "$BLOB_DIR"/test-snapd-python-webserver*fake1*.snap
+
+    echo "And a refresh is performed"
+    snap refresh --transactional test-snapd-tools test-snapd-python-webserver
+
+    echo "Then the new versions are installed"
+    for snap in test-snapd-tools test-snapd-python-webserver; do
+        snap list | MATCH "$snap .*fake1"
+    done
```
Comment (bboozzoo) [inline tests/main/refresh-many-transactional/task.yaml:65]:
```suggestion
    for name in test-snapd-tools test-snapd-python-webserver; do
        snap list "$name" | MATCH "$name .*fake1"
    done
```

File: tests/main/refresh-many-transactional/task.yaml [hunk 5]
```diff
@@ -0,0 +1,65 @@
+summary: Check that more than one snap is refreshed.
```
Comment (bboozzoo) [inline tests/main/refresh-many-transactional/task.yaml:1]:
```suggestion
summary: Check transactional refresh of more than one snap.
```

**Review by bboozzoo — COMMENTED** [review]
It'd be nice to have a managers level test for the install and refresh scenarios

File: overlord/snapstate/snapstate_install_test.go [hunk 6]
```diff
@@ -4436,6 +4491,113 @@ epoch: 1
...
+`, name)
+		paths = append(paths, makeTestSnap(c, yaml))
+		sideInfos = append(sideInfos, &snap.SideInfo{RealName: name})
+	}
+
+	s.o.TaskRunner().AddHandler("fail", func(*state.Task, *tomb.Tomb) error {
+		return errors.New("expected")
+	}, nil)
+
+	// transactional == true
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate_install_test.go:4568]: Thanks for pointing out, it was some left-over from a previous iteration.

File: overlord/snapstate/snapstate_install_test.go [hunk 7]
```diff
@@ -4436,6 +4491,113 @@ epoch: 1
...
+	c.Assert(err, IsNil)
+	c.Assert(tss, HasLen, 2)
+
+	// fail installation of 'other-snap' which will affect 'some-snap'
+	failingTask := s.state.NewTask("fail", "expected failure")
+	snapThreeLanes := tss[1].Tasks()[0].Lanes()
+	for _, lane := range snapThreeLanes {
+		failingTask.JoinLane(lane)
+	}
+	tss[1].AddTask(failingTask)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate_install_test.go:4580]: Great suggestion, thanks

File: tests/main/refresh-many-transactional-undo/task.yaml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    retry -n 4 --wait 0.5 test -e "$BLOB_DIR/${BAD_SNAP}"*fake1*.snap
+
+    echo "When a snap is broken"
+    echo "i-am-broken-now" >> "$BLOB_DIR/${BAD_SNAP}"*fake1*.snap
+
+    echo "And a refresh is performed"
+    if snap refresh --transactional "$GOOD_SNAP" "$BAD_SNAP" ; then
+        echo "snap refresh should fail but it did not, test is broken"
+        exit 1
+    fi
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply tests/main/refresh-many-transactional-undo/task.yaml:63]: Well, minus the echo, but that is a bit redundant anyway. I will change this.

File: tests/main/refresh-many-transactional-undo/task.yaml [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    echo "Good snap did not either as it was a transactional refresh"
+    snap list | MATCH -E "${GOOD_SNAP}" | NOMATCH "fake"
+
+    #shellcheck source=tests/lib/changes.sh
+    . "$TESTSLIB"/changes.sh
+    chg_id=$(change_id "Refresh snap" Error)
+
+    echo "Verify the snap change"
+    snap change "$chg_id" | MATCH "Undone.*Download snap \"${BAD_SNAP}\""
+    snap change "$chg_id" | MATCH "Undone.*Download snap \"${GOOD_SNAP}\""
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply tests/main/refresh-many-transactional-undo/task.yaml]: Great point, and actually the spread tests already caught this. Already fixed.

File: overlord/snapstate/snapstate.go [hunk 10]
```diff
@@ -1253,7 +1256,11 @@ func InstallMany(st *state.State, names []string, userID int, flags *Flags) ([]s
 		if err != nil {
 			return nil, nil, err
 		}
-		ts.JoinLane(st.NewLane())
+		if flags.Transactional {
+			ts.JoinLane(transactionLane)
+		} else {
+			ts.JoinLane(st.NewLane())
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go:1262]: Isn't this a bit redundant as we already have a similar comment a few lines above?

[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go:1262]: or maybe something in the lines of: `// look at many snap install handling for explanation how lanes are used` ?

[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go:1262]: oh, w8 that's the same function, i think this comment is actually useful, but you can replace the ones in doUpdate() maybe?

**Comment by alfonsosanchezbeato** [comment]
Pushed changes to address comments, including more manager tests.

File: overlord/snapstate/snapstate.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go:1266]: I have added something alternative in the part inside the loop, let me know if that works for you.

**Review by MiguelPires — APPROVED** [review]
LGTM, thank you

File: overlord/snapstate/snapstate.go [hunk 11]
```diff
@@ -1218,13 +1218,16 @@ func InstallMany(st *state.State, names []string, userID int, flags *Flags) ([]s
...
+		transactionLane = st.NewLane()
+	}
 	tasksets := make([]*state.TaskSet, 0, len(installs))
 	for _, sar := range installs {
 		info := sar.Info
 		var snapst SnapState
-		var flags Flags
 
-		flags, err := ensureInstallPreconditions(st, info, flags, &snapst)
+		validatedFlags, err := ensureInstallPreconditions(st, info, *flags, &snapst)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1230]: we probably want a test like TestInstallPathManyDevMode but for InstallMany to check for the flags code here

**Review by pedronis — APPROVED** [review]
thanks, one comment about testing. Please also wait for @bboozzoo re-review

**Review by bboozzoo — APPROVED** [review]
LGTM, thanks!

File: overlord/snapstate/snapstate.go (hunk 11 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go:1230]: I've pushed now a commit with a new test.

**Pull request merged**